lotman/lotman_ui.go

```diff
@@ -0,0 +1,573 @@
+//go:build linux && !ppc64le
+
+/***************************************************************
+*
+* Copyright (C) 2024, Pelican Project, Morgridge Institute for Research
+*
+* Licensed under the Apache License, Version 2.0 (the "License"); you
+* may not use this file except in compliance with the License.  You may
+* obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*
+***************************************************************/
+
+package lotman
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"unsafe"
+
+	"github.com/gin-gonic/gin"
+	"github.com/lestrrat-go/jwx/v2/jwt"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/pelicanplatform/pelican/config"
+	"github.com/pelicanplatform/pelican/param"
+	"github.com/pelicanplatform/pelican/server_utils"
+	"github.com/pelicanplatform/pelican/token"
+	"github.com/pelicanplatform/pelican/token_scopes"
+	"github.com/pelicanplatform/pelican/utils"
+)
+
+// Given a token and a list of authorized callers, check that the token is signed by one of the authorized callers. Return
+// a pointer to the parsed token.
+func tokenSignedByAuthorizedCaller(strToken string, authorizedCallers *[]string) (bool, *jwt.Token, error) {
+	ownerFound := false
+	var tok jwt.Token
+	for _, owner := range *authorizedCallers {
+		kSet, err := server_utils.GetJWKSFromIssUrl(owner)
+		if err != nil {
+			log.Debugf("Error getting JWKS for owner %s: %v", owner, err)
+			continue
+		}
+		tok, err = jwt.Parse([]byte(strToken), jwt.WithKeySet(*kSet), jwt.WithValidate(true))
+		if err != nil {
+			log.Debugf("Token verification failed with owner %s: %v -- skipping", owner, err)
+			continue
+		}
+		ownerFound = true
+		break
+	}
+
+	if !ownerFound {
+		return false, nil, errors.New("token not signed by any of the owners of any parent lot")
+	}
+
+	return true, &tok, nil
+}
+
+// Verify that a token received is a valid token. Upon verification, we set the lot's parents/owner to the
+// appropriate values. Returns true if the token is valid, false otherwise.
+func VerifyNewLotToken(lot *Lot, strToken string) (bool, error) {
+	tokenApproved := false
+
+	// Get the path associated with the lot (right now we assume/enforce a single path) and try
+	// to deduce a namespace prefix from that. We'll use that namespace prefix's issuer as the
+	// owner of the lot. If there is no associated isuer, we assign ownership to the federation
```

Comment:
Like we discussed, owner in this case is an overloaded term. Maybe call it the data-owner? Or some other name. And all data-owners are lot-managers of any child lot?